### ISSUE TITLE
> PHP 5.4 ?

### ISSUE THREAD

**Issue by bsolano** [defuse/php-encryption#277]
I am getting Uncaught exception 'Defuse\Crypto\Exception\EnvironmentIsBrokenException' with message 'Cipher method not supported.' with PHP 5.4.44 for Windows (I am stuck with that version).  The main page of the project says that the minimum requirement is PHP 5.4.

I do not know why PHP 5.4 for Windows on the openssl extension does not have aes-256-ctr.  I guess it is a Windows issue.  Can another cypher method be used?  These are the methods in the Windows version:

array(110) {
  [0]=>
  string(11) "AES-128-CBC"
  [1]=>
  string(11) "AES-128-CFB"
  [2]=>
  string(12) "AES-128-CFB1"
  [3]=>
  string(12) "AES-128-CFB8"
  [4]=>
  string(11) "AES-128-ECB"
  [5]=>
  string(11) "AES-128-OFB"
  [6]=>
  string(11) "AES-192-CBC"
  [7]=>
  string(11) "AES-192-CFB"
  [8]=>
  string(12) "AES-192-CFB1"
  [9]=>
  string(12) "AES-192-CFB8"
  [10]=>
  string(11) "AES-192-ECB"
  [11]=>
  string(11) "AES-192-OFB"
  [12]=>
  string(11) "AES-256-CBC"
  [13]=>
  string(11) "AES-256-CFB"
  [14]=>
  string(12) "AES-256-CFB1"
  [15]=>
  string(12) "AES-256-CFB8"
  [16]=>
  string(11) "AES-256-ECB"
  [17]=>
  string(11) "AES-256-OFB"
  [18]=>
  string(6) "BF-CBC"
  [19]=>
  string(6) "BF-CFB"
  [20]=>
  string(6) "BF-ECB"
  [21]=>
  string(6) "BF-OFB"
  [22]=>
  string(9) "CAST5-CBC"
  [23]=>
  string(9) "CAST5-CFB"
  [24]=>
  string(9) "CAST5-ECB"
  [25]=>
  string(9) "CAST5-OFB"
  [26]=>
  string(7) "DES-CBC"
  [27]=>
  string(7) "DES-CFB"
  [28]=>
  string(8) "DES-CFB1"
  [29]=>
  string(8) "DES-CFB8"
  [30]=>
  string(7) "DES-ECB"
  [31]=>
  string(7) "DES-EDE"
  [32]=>
  string(11) "DES-EDE-CBC"
  [33]=>
  string(11) "DES-EDE-CFB"
  [34]=>
  string(11) "DES-EDE-OFB"
  [35]=>
  string(8) "DES-EDE3"
  [36]=>
  string(12) "DES-EDE3-CBC"
  [37]=>
  string(12) "DES-EDE3-CFB"
  [38]=>
  string(13) "DES-EDE3-CFB1"
  [39]=>
  string(13) "DES-EDE3-CFB8"
  [40]=>
  string(12) "DES-EDE3-OFB"
  [41]=>
  string(7) "DES-OFB"
  [42]=>
  string(8) "DESX-CBC"
  [43]=>
  string(8) "IDEA-CBC"
  [44]=>
  string(8) "IDEA-CFB"
  [45]=>
  string(8) "IDEA-ECB"
  [46]=>
  string(8) "IDEA-OFB"
  [47]=>
  string(10) "RC2-40-CBC"
  [48]=>
  string(10) "RC2-64-CBC"
  [49]=>
  string(7) "RC2-CBC"
  [50]=>
  string(7) "RC2-CFB"
  [51]=>
  string(7) "RC2-ECB"
  [52]=>
  string(7) "RC2-OFB"
  [53]=>
  string(3) "RC4"
  [54]=>
  string(6) "RC4-40"
  [55]=>
  string(11) "aes-128-cbc"
  [56]=>
  string(11) "aes-128-cfb"
  [57]=>
  string(12) "aes-128-cfb1"
  [58]=>
  string(12) "aes-128-cfb8"
  [59]=>
  string(11) "aes-128-ecb"
  [60]=>
  string(11) "aes-128-ofb"
  [61]=>
  string(11) "aes-192-cbc"
  [62]=>
  string(11) "aes-192-cfb"
  [63]=>
  string(12) "aes-192-cfb1"
  [64]=>
  string(12) "aes-192-cfb8"
  [65]=>
  string(11) "aes-192-ecb"
  [66]=>
  string(11) "aes-192-ofb"
  [67]=>
  string(11) "aes-256-cbc"
  [68]=>
  string(11) "aes-256-cfb"
  [69]=>
  string(12) "aes-256-cfb1"
  [70]=>
  string(12) "aes-256-cfb8"
  [71]=>
  string(11) "aes-256-ecb"
  [72]=>
  string(11) "aes-256-ofb"
  [73]=>
  string(6) "bf-cbc"
  [74]=>
  string(6) "bf-cfb"
  [75]=>
  string(6) "bf-ecb"
  [76]=>
  string(6) "bf-ofb"
  [77]=>
  string(9) "cast5-cbc"
  [78]=>
  string(9) "cast5-cfb"
  [79]=>
  string(9) "cast5-ecb"
  [80]=>
  string(9) "cast5-ofb"
  [81]=>
  string(7) "des-cbc"
  [82]=>
  string(7) "des-cfb"
  [83]=>
  string(8) "des-cfb1"
  [84]=>
  string(8) "des-cfb8"
  [85]=>
  string(7) "des-ecb"
  [86]=>
  string(7) "des-ede"
  [87]=>
  string(11) "des-ede-cbc"
  [88]=>
  string(11) "des-ede-cfb"
  [89]=>
  string(11) "des-ede-ofb"
  [90]=>
  string(8) "des-ede3"
  [91]=>
  string(12) "des-ede3-cbc"
  [92]=>
  string(12) "des-ede3-cfb"
  [93]=>
  string(13) "des-ede3-cfb1"
  [94]=>
  string(13) "des-ede3-cfb8"
  [95]=>
  string(12) "des-ede3-ofb"
  [96]=>
  string(7) "des-ofb"
  [97]=>
  string(8) "desx-cbc"
  [98]=>
  string(8) "idea-cbc"
  [99]=>
  string(8) "idea-cfb"
  [100]=>
  string(8) "idea-ecb"
  [101]=>
  string(8) "idea-ofb"
  [102]=>
  string(10) "rc2-40-cbc"
  [103]=>
  string(10) "rc2-64-cbc"
  [104]=>
  string(7) "rc2-cbc"
  [105]=>
  string(7) "rc2-cfb"
  [106]=>
  string(7) "rc2-ecb"
  [107]=>
  string(7) "rc2-ofb"
  [108]=>
  string(3) "rc4"
  [109]=>
  string(6) "rc4-40"
}

**Comment by sstok** [comment]
You can try updating the libssl.dll and liblay.dll to see if this works, PHP 5.4 is very old and EOL so the OpenSSL version shipped with that version is outdated as well.

**Comment by paragonie-scott** [comment]
> I do not know why PHP 5.4 for Windows on the openssl extension does not have aes-256-ctr. I guess it is a Windows issue.

I suspect what @sstok says is likely accurate: It's probably compiled with an outdate OpenSSL. Alternatively, it's FIPS mode OpenSSL which is severely crippled.

> Can another cypher method be used? These are the methods in the Windows version.

Not without forking the entire project and reintroducing CBC mode.

**Comment by bsolano** [comment]
@sstok, thank you.  I will try to change those DLLs.  Although if I am not wrong it should not be that straightforward because the signature of the DLLs could be different with what was used at compilation.  But I'll try anyway.

@paragonie-scott in fact I just tried without knowing very much what I was doing to change the CIPHER_METHOD in Core and it did not work.  Tests throws an exception in AESTestVector().  

> Try 5.6 maybe?

Just to clarify, I am stuck with PHP 5.4.44 because I am stuck with a Windows 2003 server box.  That windows version supposedly only runs PHP compiled with VC9.  PHP > 5.5 is compiled with newer versions of Visual Studio.  Maybe, that's the reason PHP 5.4.44 openssl extension is compiled with an old version of openssl, because they had to compile it against an openssl compiled with VC9.  I tested your library with PHP 5.5 on Windows and it works just fine.  Maybe README.md should say that > 5.4 but for Windows > 5.5.

I have compiled PHP in Windows in the past, but is not as easy as with Unix like systems.  Although that was a long time ago. I guess that would be a last resort.

A last question: if I use legacy methods, runtimeTest() will still be run?  Is that so?  I think it is from a shallow look at the code.

Thank you guys for your support and your work!

**Comment by paragonie-scott** [comment]
The runtime tests were written based on test vectors published by NIST for AES-256-CTR. If you want to fork the library to work with another ciphersuite, you're going to have a difficult time.

What's stopping you from migrating the application to a newer Windows server with PHP 5.6 support? 5.4 is dead in the water.

**Comment by defuse** [comment]
@paragonie-scott thanks for chiming in. Yeah, switching away from CTR mode would require a significant rewrite of the file encryption code.

**Comment by paragonie-scott** [comment]
Do we detect if `aes-256-ctr` exists at runtime? If not, we could throw an exception stating "Outdated version of OpenSSL or FIPS mode detected. Cannot proceed."

**Comment by defuse** [comment]
We could change the "Cipher method not supported." error message to that.

**Comment by paragonie-scott** [comment]
Sure, I'll send a PR.

**Comment by sstok** [comment]
"I have compiled PHP in Windows in the past, but is not as easy as with Unix like systems."
Yea I know 😄  not to mention that installing visual studio is like unleashing the blob on your system, everything gets affected by it. I did it once for Zend Studio which was missing the Intl extension, and it was no fun 💀

aes-256-ctr is supported since OpenSSL v1.0.1. So if PHP 5.4 supports v1.0.x then any higher version should work.

I found https://anaconda.org/anaconda/openssl/1.0.2h/download/win-64/openssl-1.0.2h-vc9_0.tar.bz2 which seems to be the latest version with VS9! and find the OpenSSL API version of PHP-5.4 so I'm not sure it works.
